File: core/services/performance/fsp_tracing/fsp_tracer.cc
// Copyright 2025 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "core/services/performance/fsp_tracing/fsp_tracer.h"

#include <cmath>
#include <cstdint>
#include <string>
#include <utility>

#include "base/trace/native/trace_event.h"
#include "core/public/performance_controller_platform_impl.h"
#include "core/services/performance/performance_controller.h"
#include "core/services/timing_handler/timing_constants.h"
#include "core/services/trace/service_trace_event_def.h"

namespace lynx {
namespace tasm {
namespace performance {

void FSPTracer::Start(CompletionCallback completion_callback) {
  if (!config_.enable_ || is_running_) {
    return;
  }
  completion_callback_ = std::move(completion_callback);
  TRACE_EVENT(LYNX_TRACE_CATEGORY, "FSPTracer::Start");
  is_running_ = true;
  // Start hard timeout
  if (config_.hard_timeout_ms_ <= 0) {
    return;
  }
  std::weak_ptr<FSPTracer> weak_this = shared_from_this();
  PerformanceController::GetTaskRunner()->PostDelayedTask(
      [weak_this]() {
        auto this_ptr = weak_this.lock();
        if (this_ptr && this_ptr->is_running_) {
          this_ptr->is_running_ = false;
          this_ptr->OnFSPHardTimeOut(this_ptr->previous_snapshot_,
                                     base::CurrentSystemTimeMicroseconds());
        }
      },
      fml::TimeDelta::FromMilliseconds(config_.hard_timeout_ms_));
}

void FSPTracer::Stop(int64_t current_timestamp_us) {
  if (!config_.enable_ || !is_running_) {
    return;
  }
  is_running_ = false;
  OnFSPStop(previous_snapshot_, current_timestamp_us);
}

void FSPTracer::CancelledByUserInteraction(int64_t current_timestamp_us) {
  if (!config_.enable_ || !is_running_) {
    return;
  }
  is_running_ = false;
  // Send cancelled event
  OnFSPCancelledByUserInteraction(previous_snapshot_, current_timestamp_us);
}

void FSPTracer::OnCaptureSnapshot(FSPSnapshot current_snapshot) {
  if (!config_.enable_ || !is_running_) {
    return;
  }
  // 1. Check if the snapshot is valuable.
  if (!IsSnapshotValuable(current_snapshot, config_)) {
    // If not satisfied, it proves instability, so the previous snapshot is also
    // invalid
    previous_snapshot_.reset();
    return;
  }
  TRACE_EVENT_INSTANT(
      LYNX_TRACE_CATEGORY_VITALS, FSP_TRACER_ON_VALUABLE_SNAPSHOT,
      [&current_snapshot,
       instance_id = instance_id_](lynx::perfetto::EventContext ctx) {
        ctx.event()->set_timestamp_absolute_us(
            current_snapshot.trace_timestamp_us_);
        ctx.event()->add_debug_annotations(INSTANCE_ID,
                                           std::to_string(instance_id));
        ctx.event()->add_debug_annotations(
            FSP_CONTAINER_SIZE,
            std::to_string(current_snapshot.container_size_.Width() *
                           current_snapshot.container_size_.Height()));
        ctx.event()->add_debug_annotations(
            FSP_TOTAL_PRESENTED_CONTENT_AREA,
            std::to_string(current_snapshot.total_presented_content_area_));
        ctx.event()->add_debug_annotations(
            timing::kContentFillPercentageX,
            std::to_string(current_snapshot.content_fill_percentage_x_));
        ctx.event()->add_debug_annotations(
            timing::kContentFillPercentageY,
            std::to_string(current_snapshot.content_fill_percentage_y_));
        ctx.event()->add_debug_annotations(
            timing::kContentFillPercentageTotalArea,
            std::to_string(
                current_snapshot.content_fill_percentage_total_area_));
        ctx.event()->add_debug_annotations(
            timing::kContainerFillPercentageContainerArea,
            std::to_string(
                current_snapshot.container_fill_percentage_container_area_));
        ctx.event()->add_debug_annotations(
            FSP_LAST_CHANGE_TIMESTAMP_US,
            std::to_string(current_snapshot.last_change_timestamp_us_));
      });
  auto& previous_snapshot = previous_snapshot_;
  // 2. If previous snapshot exists (timestamp > 0), check if the snapshot is
  // stable.
  if (previous_snapshot.has_value() &&
      previous_snapshot->last_change_timestamp_us_ > 0 &&
      IsSnapshotStable(current_snapshot, *previous_snapshot, config_)) {
    TRACE_EVENT(LYNX_TRACE_CATEGORY, FSP_TRACER_SNAPSHOT_STABLE);
    int64_t diff_t_ms = (current_snapshot.last_change_timestamp_us_ -
                         previous_snapshot->last_change_timestamp_us_) /
                        1000;
    // If snapshot is stable, check if the interval is less than the min
    // interval.
    if (diff_t_ms >= config_.min_diff_interval_ms_) {
      // If the interval is greater than the min interval, generate FSP.
      // Generate FSP, previous_snapshot is the snapshot at FSP time
      // Construct FSPEntry based on previous_snapshot and send it out.
      OnFSP(previous_snapshot);
    }
    return;
  }
  TRACE_EVENT(LYNX_TRACE_CATEGORY, FSP_TRACER_UPDATE_PREVIOUS_SNAPSHOT_STABLE);
  // 3. Update previous snapshot, if the snapshot is valuable.
  previous_snapshot_ = std::move(current_snapshot);
}

bool FSPTracer::IsSnapshotValuable(FSPSnapshot& snapshot,
                                   const FSPConfig& config) const {
  // 1. Check X and Y axis projection fill rates
  auto total_projection_w = snapshot.x_total_content_projections_.count();
  if (total_projection_w <= 0) {
    return false;
  }
  snapshot.content_fill_percentage_x_ = static_cast<int32_t>(
      snapshot.x_projections_.count() * 100 / total_projection_w);
  if (snapshot.content_fill_percentage_x_ <
      config.min_content_fill_percentage_x_) {
    return false;
  }

  // 2. Check effective area fill rate
  auto total_projection_h = snapshot.y_total_content_projections_.count();
  if (total_projection_h <= 0) {
    return false;
  }
  snapshot.content_fill_percentage_y_ = static_cast<int32_t>(
      (snapshot.y_projections_.count() * 100 / total_projection_h));
  if (snapshot.content_fill_percentage_y_ <
      config.min_content_fill_percentage_y_) {
    return false;
  }

  // 3. Check effective area fill rate
  if (snapshot.total_content_area_ <= 0) {
    return false;
  }

  // 4. Check meaningful content presented rate
  snapshot.content_fill_percentage_total_area_ =
      static_cast<int32_t>(snapshot.total_presented_content_area_ * 100 /
                           snapshot.total_content_area_);
  if (snapshot.content_fill_percentage_total_area_ <
      config.min_content_fill_percentage_total_area_) {
    return false;
  }

  // 5. Check container fill rate
  auto container_area =
      snapshot.container_size_.Width() * snapshot.container_size_.Height();
  if (container_area <= 0) {
    return false;
  }
  snapshot.container_fill_percentage_container_area_ = static_cast<int32_t>(
      snapshot.total_presented_content_area_ * 100 / container_area);
  return snapshot.container_fill_percentage_container_area_ >=
         config.min_container_fill_percentage_container_area_;
}

bool FSPTracer::IsSnapshotStable(const FSPSnapshot& current,
                                 const FSPSnapshot& previous,
                                 const FSPConfig& config) {
  TRACE_EVENT(LYNX_TRACE_CATEGORY, "FSPTracer::IsSnapshotStable");
  if (previous.last_change_timestamp_us_ <= 0 ||
      current.last_change_timestamp_us_ <= 0) {
    // Initial snapshot always considered stable.
    return false;
  }
  // diff interval in milliseconds
  int64_t diff_t_ms =
      (current.last_change_timestamp_us_ - previous.last_change_timestamp_us_) /
      1000;
  if (diff_t_ms <= 0) {
    return false;
  }
  // 1. (fast) Projection change rate
  int acceptable_pixel_diff_per_sec = config.acceptable_pixel_diff_per_sec_;
  int change_rate_w =
      std::abs(static_cast<int>(
          (current.x_projections_.count() - previous.x_projections_.count()))) *
      1000 / diff_t_ms;
  if (change_rate_w > acceptable_pixel_diff_per_sec) {
    return false;
  }
  int change_rate_h =
      std::abs(static_cast<int>(
          (current.y_projections_.count() - previous.y_projections_.count()))) *
      1000 / diff_t_ms;
  if (change_rate_h > acceptable_pixel_diff_per_sec) {
    return false;
  }

  // 2.1 (fast) Area projection change rate
  int area_change_rate_w =
      std::abs(static_cast<int>((current.total_presented_content_area_ -
                                 previous.total_presented_content_area_))) *
      1000 / diff_t_ms;
  if (area_change_rate_w > config.acceptable_area_diff_per_sec_) {
    return false;
  }

  // 2.2 (fast) Container fill rate change rate
  int container_fill_rate_change_rate_w =
      std::abs(static_cast<int>(
          (current.container_fill_percentage_container_area_ -
           previous.container_fill_percentage_container_area_))) *
      1000 / diff_t_ms;
  if (container_fill_rate_change_rate_w >
      config.acceptable_area_diff_per_sec_) {
    return false;
  }

  // 3. (slow) XOR projection change rate
  auto xor_change_rate_x =
      static_cast<int>(
          (current.x_projections_ ^ previous.x_projections_).count()) *
      1000 / diff_t_ms;
  if (xor_change_rate_x > acceptable_pixel_diff_per_sec) {
    return false;
  }
  auto xor_change_rate_y =
      static_cast<int>(
          (current.y_projections_ ^ previous.y_projections_).count()) *
      1000 / diff_t_ms;
  return (xor_change_rate_y <= acceptable_pixel_diff_per_sec);
}

void FSPTracer::OnFSP(const base::flex_optional<FSPSnapshot>& fsp_snapshot) {
  is_running_ = false;
  HandleFSPResult((fsp_snapshot.has_value() ? FSPResult::kFSPSuccess
                                            : FSPResult::kFSPError),
                  fsp_snapshot, -1);
}

void FSPTracer::OnFSPStop(
    const base::flex_optional<FSPSnapshot>& previous_snapshot,
    int64_t current_timestamp_us) {
  HandleFSPResult(FSPResult::kFSPStop, previous_snapshot, current_timestamp_us);
}

void FSPTracer::OnFSPHardTimeOut(
    const base::flex_optional<FSPSnapshot>& current_snapshot,
    int64_t current_timestamp_us) {
  HandleFSPResult(FSPResult::kFSPCancelByTimeout, current_snapshot,
                  current_timestamp_us);
}

void FSPTracer::OnFSPCancelledByUserInteraction(
    const base::flex_optional<FSPSnapshot>& current_snapshot,
    int64_t current_timestamp_us) {
  HandleFSPResult(FSPResult::kFSPCancelByUserInteraction, current_snapshot,
                  current_timestamp_us);
}

void FSPTracer::HandleFSPResult(
    const char* status,
    const base::flex_optional<FSPSnapshot>& current_snapshot,
    int64_t current_timestamp_us) {
  if (!completion_callback_) {
    return;
  }
  TRACE_EVENT_INSTANT(
      LYNX_TRACE_CATEGORY_VITALS, FSP_TIMING_MARK_FSP_END,
      [status, &current_snapshot,
       instance_id = instance_id_](lynx::perfetto::EventContext ctx) {
        ctx.event()->add_debug_annotations(timing::kFSPStatus, status);
        if (current_snapshot.has_value()) {
          ctx.event()->add_debug_annotations(INSTANCE_ID,
                                             std::to_string(instance_id));
          ctx.event()->add_debug_annotations(
              timing::kFSPEnd,
              std::to_string(current_snapshot->last_change_timestamp_us_));
          // Add 1 us to make sure this evnet is after the snapshot bitmap
          // event.
          ctx.event()->set_timestamp_absolute_us(
              current_snapshot->trace_timestamp_us_ + 1);
          ctx.event()->set_track_uuid(current_snapshot->trace_thread_id_);
        }
      });
  if (!current_snapshot.has_value()) {
    completion_callback_(FSPResult(status, current_timestamp_us));
    return;
  }
  FSPResult result(status, current_snapshot->last_change_timestamp_us_,
                   current_snapshot->content_fill_percentage_x_,
                   current_snapshot->content_fill_percentage_y_,
                   current_snapshot->content_fill_percentage_total_area_,
                   current_snapshot->container_fill_percentage_container_area_);
  completion_callback_(std::move(result));
}

}  // namespace performance
}  // namespace tasm
}  // namespace lynx

File: platform/android/lynx_android/src/main/java/com/lynx/tasm/performance/fsp/FSPTracer.java
// Copyright 2025 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.tasm.performance.fsp;

import android.graphics.Rect;
import androidx.annotation.AnyThread;
import androidx.annotation.UiThread;
import com.lynx.tasm.base.TraceEvent;
import com.lynx.tasm.base.trace.TraceEventDef;
import com.lynx.tasm.behavior.ui.ILynxUIMeaningfulContent;
import com.lynx.tasm.behavior.ui.MeaningfulPaintingArea;
import com.lynx.tasm.eventreport.LynxEventReporter;
import com.lynx.tasm.performance.PerformanceController;
import com.lynx.tasm.utils.UIThreadUtils;
import java.lang.ref.WeakReference;
import java.util.BitSet;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * FSP (First Stable Paint) Tracer implementation for Android platform.
 * Based on C++ core implementation and iOS interface.
 */
public class FSPTracer {
  // Status enum for type-safe state representation
  public enum ResultStatus {
    SUCCESS("success"),
    CANCEL_BY_USER_INTERACTION("cancelByUserInteraction"),
    CANCEL_BY_TIMEOUT("timeout"),
    STOP("stop"),
    ERROR("error");

    private final String value;

    ResultStatus(String value) {
      this.value = value;
    }

    public String getValue() {
      return value;
    }

    // Helper method to get enum from string value
    public static ResultStatus fromValue(String value) {
      for (ResultStatus status : ResultStatus.values()) {
        if (status.value.equals(value)) {
          return status;
        }
      }
      return null;
    }
  }

  private static final String TAG = "FSPTracer";
  // HashMap keys for FSP timing info
  private static final String KEY_FSP_STATUS = "fspStatus";
  private static final String KEY_CONTENT_FILL_PERCENTAGE_X = "contentFillPercentageX";
  private static final String KEY_CONTENT_FILL_PERCENTAGE_Y = "contentFillPercentageY";
  private static final String KEY_CONTENT_FILL_PERCENTAGE_TOTAL_AREA =
      "contentFillPercentageTotalArea";
  private static final String KEY_CONTAINER_FILL_PERCENTAGE_CONTAINER_AREA =
      "containerFillPercentageContainerArea";
  // Tracer fields
  private final FSPConfig mConfig = new FSPConfig();
  private final AtomicBoolean mIsRunning = new AtomicBoolean(false);
  private volatile FSPSnapshot mPreviousSnapshot;
  private WeakReference<PerformanceController> mPerfControllerRef = new WeakReference<>(null);

  /**
   * Constructor with configuration
   * @param config FSP configuration
   */
  public FSPTracer(PerformanceController perfController) {
    if (perfController != null) {
      mPerfControllerRef = new WeakReference<>(perfController);
    }
  }

  /**
   * Start monitoring for FSP
   * @param callback Completion callback to be invoked when FSP is detected
   */
  @UiThread
  public void start(IMeaningfulContentSnapshotCaptureHandler captureHandler) {
    if (captureHandler == null || !mConfig.enable || mIsRunning.get()) {
      return;
    }
    // 1. setup config.
    mConfig.parse();
    mIsRunning.set(true);
    // 2. start snapshot timer
    scheduleNextCapture(captureHandler);
    // 3. Start hard timeout if configured
    startHardTimeout();
  }

  /**
   * Stop monitoring
   * @param currentTimestampUs Current timestamp in microseconds
   */
  @UiThread
  public void stop() {
    if (!mConfig.enable || !mIsRunning.get()) {
      return;
    }
    internalStop(ResultStatus.STOP);
  }

  /**
   * Cancel monitoring due to user interaction
   * @param currentTimestampUs Current timestamp in microseconds
   */
  @UiThread
  public void cancelledByUserInteraction() {
    if (!mConfig.enable || !mIsRunning.get()) {
      return;
    }
    internalStop(ResultStatus.CANCEL_BY_USER_INTERACTION);
  }

  @UiThread
  private void scheduleNextCapture(final IMeaningfulContentSnapshotCaptureHandler captureHandler) {
    if (mConfig.snapshotIntervalMs <= 0 || captureHandler == null) {
      return;
    }
    // Use a weak reference to prevent memory leaks
    final WeakReference<FSPTracer> weakSelf = new WeakReference<>(this);
    UIThreadUtils.runOnUiThread(() -> {
      FSPTracer tracer = weakSelf.get();
      if (tracer == null || !tracer.mIsRunning.get() || captureHandler == null) {
        return;
      }
      // Capture the snapshot
      MeaningfulContentSnapshot snapshot = captureHandler.capture();
      if (snapshot == null) {
        // Schedule next capture even if current capture returns null
        tracer.scheduleNextCapture(captureHandler);
        return;
      }
      tracer.handleSnapshotCapture(snapshot);
      // For now, we'll just schedule the next capture
      // In a real implementation, we would convert the snapshot and process it
      tracer.scheduleNextCapture(captureHandler);
    }, mConfig.snapshotIntervalMs);
  }

  @UiThread
  private void startHardTimeout() {
    if (mConfig.hardTimeoutMs <= 0) {
      return;
    }
    LynxEventReporter.delayRunOnReportThread(() -> {
      if (!mConfig.enable || !mIsRunning.get()) {
        return;
      }
      internalStop(ResultStatus.CANCEL_BY_TIMEOUT);
    }, mConfig.hardTimeoutMs);
  }

  @AnyThread
  private void internalStop(ResultStatus status) {
    mIsRunning.set(false);
    long currentTimestampUs = PerformanceController.currentSystemTimeMicroseconds();
    if (currentTimestampUs <= 0) {
      return;
    }
    handleFSPResult(status, mPreviousSnapshot, currentTimestampUs);
  }

  @UiThread
  /// Process raw meaningful content snapshot to fsp snapshot, and check if it is valuable and
  /// stable.
  private void handleSnapshotCapture(final MeaningfulContentSnapshot rawSnapshot) {
    if (rawSnapshot.getContainerWidth() <= 0 || rawSnapshot.getContainerHeight() <= 0
        || rawSnapshot.getMeaningfulPaintingAreas() == null
        || rawSnapshot.getMeaningfulPaintingAreas().isEmpty()) {
      return;
    }
    long currentTimestampUs = PerformanceController.currentSystemTimeMicroseconds();
    final WeakReference<FSPTracer> weakSelf = new WeakReference<>(this);
    LynxEventReporter.runOnReportThread(() -> {
      FSPTracer tracer = weakSelf.get();
      if (tracer == null) {
        return;
      }
      List<MeaningfulPaintingArea> areaList = rawSnapshot.getMeaningfulPaintingAreas();
      if (areaList == null) {
        return;
      }
      FSPSnapshot snapshot = new FSPSnapshot(
          rawSnapshot.getContainerWidth(), rawSnapshot.getContainerHeight(), currentTimestampUs);
      for (MeaningfulPaintingArea area : areaList) {
        boolean isPresented = area.getMeaningfulContentStatus()
            == ILynxUIMeaningfulContent.MeaningfulContentStatus.PRESENTED;

        snapshot.fillContentToSnapshot(isPresented,
            new Rect(area.getOffsetX(), area.getOffsetY(), area.getOffsetX() + area.getWidth(),
                area.getOffsetY() + area.getHeight()),
            area.getFirstMeaningfulContentPresentedTimestampMicros());
      }
      snapshot.traceCurrentTimestampUs = rawSnapshot.getTraceCurrentTimestampUs();
      tracer.onCaptureSnapshot(snapshot);
    });
  }

  /// @note Run on ReportThread
  /// Process captured fsp snapshot
  public void onCaptureSnapshot(FSPSnapshot snapshot) {
    if (!mConfig.enable || !mIsRunning.get()) {
      return;
    }

    // Check if the snapshot is valuable
    if (!isSnapshotValuable(snapshot, mConfig)) {
      // If not valuable, reset previous snapshot
      mPreviousSnapshot = null;
      return;
    }
    if (TraceEvent.isTracingStarted()) {
      TraceEvent.instant(TraceEventDef.CATEGORY_DEFAULT,
          TraceEventDef.FSP_TRACER_ON_VALUABLE_SNAPSHOT, snapshot.traceCurrentTimestampUs,
          snapshot.toMap());
    }
    // If previous snapshot exists, check stability
    if (mPreviousSnapshot != null && mPreviousSnapshot.getLastChangeTimestampUs() > 0
        && isSnapshotStable(snapshot, mPreviousSnapshot, mConfig)) {
      TraceEvent.beginSection(TraceEventDef.FSP_TRACER_SNAPSHOT_STABLE);
      long diffTUs =
          snapshot.getLastChangeTimestampUs() - mPreviousSnapshot.getLastChangeTimestampUs();
      long diffTMs = diffTUs / 1000;
      // Check if interval is greater than min interval
      if (diffTMs >= mConfig.minDiffIntervalMs) {
        // Generate FSP with previous snapshot
        mIsRunning.set(false);
        handleFSPResult(
            ResultStatus.SUCCESS, mPreviousSnapshot, mPreviousSnapshot.getLastChangeTimestampUs());
      }
      TraceEvent.endSection(TraceEventDef.FSP_TRACER_SNAPSHOT_STABLE);
      return;
    }

    // Update previous snapshot if valuable.
    mPreviousSnapshot = snapshot;
  }

  /// @note Run on ReportThread
  /// Check if snapshot is valuable for FSP calculation
  public boolean isSnapshotValuable(FSPSnapshot snapshot, FSPConfig config) {
    // Check X axis projection fill rate
    int totalProjectionW = snapshot.getXTotalContentProjections().cardinality();
    if (totalProjectionW <= 0) {
      return false;
    }

    int xProjectionCount = snapshot.getXProjections().cardinality();
    snapshot.setContentFillPercentageX(xProjectionCount * 100 / totalProjectionW);
    if (snapshot.getContentFillPercentageX() < config.minContentFillPercentageX) {
      return false;
    }

    // Check Y axis projection fill rate
    int totalProjectionH = snapshot.getYTotalContentProjections().cardinality();
    if (totalProjectionH <= 0) {
      return false;
    }

    int yProjectionCount = snapshot.getYProjections().cardinality();
    snapshot.setContentFillPercentageY(yProjectionCount * 100 / totalProjectionH);

    if (snapshot.getContentFillPercentageY() < config.minContentFillPercentageY) {
      return false;
    }

    // Check total area fill rate
    if (snapshot.getTotalContentArea() <= 0) {
      return false;
    }

    snapshot.setContentFillPercentageTotalArea(
        (int) (snapshot.getTotalPresentedContentArea() * 100 / snapshot.getTotalContentArea()));
    if (snapshot.getContentFillPercentageTotalArea() < config.minContentFillPercentageTotalArea) {
      return false;
    }

    snapshot.setContainerFillPercentageContainerArea(
        (int) (snapshot.getTotalPresentedContentArea() * 100 / snapshot.getContainerArea()));
    return snapshot.getContainerFillPercentageContainerArea()
        >= config.minContainerFillPercentageContainerArea;
  }

  /// @note Run on ReportThread
  /// Check if snapshot is stable compared to previous snapshot
  public boolean isSnapshotStable(FSPSnapshot current, FSPSnapshot previous, FSPConfig config) {
    if (previous.getLastChangeTimestampUs() <= 0 || current.getLastChangeTimestampUs() <= 0) {
      return false;
    }

    // Calculate time difference in milliseconds
    long diffTUs = current.getLastChangeTimestampUs() - previous.getLastChangeTimestampUs();
    long diffTMs = diffTUs / 1000;

    if (diffTMs <= 0) {
      return false;
    }

    // Check X projection change rate
    int changeRateW =
        Math.abs(current.getXProjections().cardinality() - previous.getXProjections().cardinality())
        * 1000 / (int) diffTMs;
    if (changeRateW > config.acceptablePixelDiffPerSec) {
      return false;
    }

    // Check Y projection change rate
    int changeRateH =
        Math.abs(current.getYProjections().cardinality() - previous.getYProjections().cardinality())
        * 1000 / (int) diffTMs;
    if (changeRateH > config.acceptablePixelDiffPerSec) {
      return false;
    }

    // Check total meaningful content area change rate
    int areaChangeRate = (int) Math.abs(current.getTotalPresentedContentArea()
                             - previous.getTotalPresentedContentArea())
        * 1000 / (int) diffTMs;
    if (areaChangeRate > config.acceptableAreaDiffPerSec) {
      return false;
    }

    // Check container area change rate
    int containerAreaChangeRate = (int) Math.abs(current.getContainerFillPercentageContainerArea()
                                      - previous.getContainerFillPercentageContainerArea())
        * 1000 / (int) diffTMs;
    if (containerAreaChangeRate > config.acceptableAreaDiffPerSec) {
      return false;
    }

    // Check XOR projection change rate
    BitSet xorX = (BitSet) current.getXProjections().clone();
    xorX.xor(previous.getXProjections());
    int xorChangeRateX = xorX.cardinality() * 1000 / (int) diffTMs;
    if (xorChangeRateX > config.acceptablePixelDiffPerSec) {
      return false;
    }

    BitSet xorY = (BitSet) current.getYProjections().clone();
    xorY.xor(previous.getYProjections());
    int xorChangeRateY = xorY.cardinality() * 1000 / (int) diffTMs;

    return xorChangeRateY <= config.acceptablePixelDiffPerSec;
  }

  @AnyThread
  /// Process fsp result and report to performance controller
  private void handleFSPResult(
      final ResultStatus status, final FSPSnapshot currentSnapshot, final long currentTimestampUs) {
    PerformanceController perfController = mPerfControllerRef.get();
    if (perfController == null) {
      return;
    }
    if (currentSnapshot == null) {
      HashMap<String, String> info = new HashMap<>(1);
      info.put(KEY_FSP_STATUS, status.getValue());
      perfController.setFSPTimingInfo(currentTimestampUs, info);
      if (TraceEvent.isTracingStarted()) {
        UIThreadUtils.runOnUiThread(() -> {
          TraceEvent.instant(
              TraceEventDef.CATEGORY_DEFAULT, TraceEventDef.FSP_TIMING_MARK_FSP_END, info);
        });
      }
      return;
    }
    HashMap<String, String> info = new HashMap<>(1);
    info.put(KEY_FSP_STATUS, status.getValue());
    info.put(
        KEY_CONTENT_FILL_PERCENTAGE_X, String.valueOf(currentSnapshot.getContentFillPercentageX()));
    info.put(
        KEY_CONTENT_FILL_PERCENTAGE_Y, String.valueOf(currentSnapshot.getContentFillPercentageY()));
    info.put(KEY_CONTENT_FILL_PERCENTAGE_TOTAL_AREA,
        String.valueOf(currentSnapshot.getContentFillPercentageTotalArea()));
    info.put(KEY_CONTAINER_FILL_PERCENTAGE_CONTAINER_AREA,
        String.valueOf(currentSnapshot.getContainerFillPercentageContainerArea()));
    perfController.setFSPTimingInfo(currentSnapshot.getLastChangeTimestampUs(), info);
    if (TraceEvent.isTracingStarted()) {
      UIThreadUtils.runOnUiThread(() -> {
        // Add 1 us to make sure this evnet is after the snapshot bitmap
        // event.
        TraceEvent.instant(TraceEventDef.CATEGORY_DEFAULT, TraceEventDef.FSP_TIMING_MARK_FSP_END,
            currentSnapshot.traceCurrentTimestampUs + 1, info);
      });
    }
  }
}
